service/history/workflow/metrics.go

```diff
@@ -131,3 +134,74 @@ func GetPerTaskQueueFamilyScope(
 		tags...,
 	)
 }
+
+type ActivityExecutionState int
+
+const (
+	ActivityStateUnknown ActivityExecutionState = iota
+	ActivityStateSucceeded
+	ActivityStateFailed
+	ActivityStateCanceled
+	ActivityStateTimeout
+)
+
+type ActivityCompletionMetrics struct {
+	// State determines whether the activity succeeded, and whether it is/will be retried
+	State ActivityExecutionState
+	// AttemptStartedTime is the start time of the current attempt
+	AttemptStartedTime time.Time
+	// FirstScheduledTime is the scheduled time of the first attempt
+	FirstScheduledTime time.Time
+	// Closed is true if no more attempts will be made to execute the activity.
+	Closed bool
+	// TimerType is the type of timer that caused the activity execution to timeout.
+	TimerType enumspb.TimeoutType
+}
+
+func RecordActivityCompletionMetrics(
+	shard historyi.ShardContext,
+	namespaceName namespace.Name,
+	taskQueue string,
+	metricsState ActivityCompletionMetrics,
+	tags ...metrics.Tag,
+) {
+	metricsHandler := GetPerTaskQueueFamilyScope(
+		shard.GetMetricsHandler(),
+		namespaceName,
+		taskQueue,
+		shard.GetConfig(),
+		tags...,
+	)
+
+	if !metricsState.AttemptStartedTime.IsZero() {
+		latency := time.Since(metricsState.AttemptStartedTime)
+		// ActivityE2ELatency is deprecated due to its inaccurate naming. It captures the attempt duration instead of an end-to-end duration as its name suggests. For now record both metrics
+		metrics.ActivityE2ELatency.With(metricsHandler).Record(latency)
+		metrics.ActivityStartToCloseLatency.With(metricsHandler).Record(latency)
+	}
```

Comment:
so start_to_close latency will be emitted in the case of timeout as well? That doesn't feel right to me as it's basically the timeout value and not the actual start to close latency from user side.